Compile error, forward-mode jacobian

Hello, I am trying to run the following code:

```
template<typename scalar_type>
Eigen::Matrix<scalar_type, 3, 3> calc_DwxpDw_wave_forward(const Eigen::Quaternion<scalar_type>& q,
                                                          const Eigen::Matrix<scalar_type, 3, 1>& p) {
    const wave::RelativeRotation<Eigen::Matrix<scalar_type, 3, 1>> w(vec_from_quat(q));
    const wave::Translation<Eigen::Matrix<scalar_type, 3, 1>> t(p);

    return (wave::exp(w) * t).jacobian(w);
}
```
Here is vec_from_quat:
```
template <typename scalar_type>
inline Eigen::Matrix<scalar_type, 3, 1> vec_from_quat( const Eigen::Quaternion<scalar_type>& q ) {
    Eigen::AngleAxis<scalar_type> aa(q);
    return aa.axis()*aa.angle();
}
```

The forward mode does not compile. I get the following console output:
```
ERROR: /srv/build-deps/toolchains/ccache/ccache-3.7.6/bin/ccache /srv/build-deps/toolchains/gcc/gcc-7.3.0/bin/g++ -g -O2 -Wall -Wextra -pthread -m64 -pipe -fexceptions -fvisibility=default -fdiagnostics-color=always -Wall -Wextra -Werror -O0 -fPIC -DNDEBUG -D_GNU_SOURCE -DPD_QBS_BUILD -Dgsl_CONFIG_CONTRACT_LEVEL_OFF -DLINUX -DLINUX_HOST -D_FILE_OFFSET_BITS=64 -DEIGEN_MPL2_ONLY 
-I$REPO/extern/wave_geometry/include 
-I$REPO/extern/wave_geometry/include/src 
-I$REPO/extern/wave_geometry/3rd-party/Tick 
-I$REPO/extern/googletest/include 
-I$REPO/extern/eigen3 -std=c++17 -o $REPO/_build/host/host_gcc/release/gradientsTest.f8af41dd/3a52ce780950d4d9/gradientsTest.cpp.o -c $REPO/test/gradientsTest.cpp

ERROR: In file included from $REPO/extern/wave_geometry/include/wave/geometry/core.hpp:45:0,
from $REPO/extern/wave_geometry/include/wave/geometry/geometry.hpp:10,
from $REPO/test/gradientsTestutils.h:14,
from $REPO/test/gradientsTest.cpp:7:
$REPO/extern/wave_geometry/include/wave/geometry/src/core/functions/TypedJacobianEvaluator.hpp: In instantiation of ‘wave::internal::jacobian_t<Derived, Target> wave::internal::evaluateTypedJacobian(const wave::ExpressionBase<T>&, const TargetDerived&) [with Derived = wave::Rotate<wave::ExpMap<const wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >&>, const wave::Translation<Eigen::Matrix<double, 3, 1> >&>; TargetDerived = wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >; typename std::enable_if<wave::internal::contains_same_type<Derived, Target>::value, int>::type <anonymous> = 0; wave::internal::jacobian_t<Derived, Target> = Eigen::Matrix<double, 3, 3>; typename std::remove_const<typename std::remove_reference<decltype (declval<typename wave::internal::traits<WrtDerived, void>::OutputFunctor>()(declval<typename wave::internal::traits<typename std::remove_const<typename std::remove_reference<typename wave::internal::traits<WrtDerived, void>::EvalType>::type>::type, void>::PlainType>()))>::type>::type = wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >; typename std::remove_const<typename std::remove_reference<decltype (declval<typename wave::internal::traits<T, void>::OutputFunctor>()(declval<typename wave::internal::traits<typename std::remove_const<typename std::remove_reference<typename wave::internal::traits<T, void>::EvalType>::type>::type, void>::PlainType>()))>::type>::type = wave::Translation<Eigen::Matrix<double, 3, 1> >; typename wave::internal::traits<typename std::remove_const<typename std::remove_reference<decltype (declval<typename wave::internal::traits<T, void>::OutputFunctor>()(declval<typename wave::internal::traits<typename std::remove_const<typename std::remove_reference<typename wave::internal::traits<T, void>::EvalType>::type>::type, void>::PlainType>()))>::type>::type>::Scalar = double]’:
$REPO/extern/wave_geometry/include/wave/geometry/src/core/functions/TypedJacobianEvaluator.hpp:314:33:   required from ‘wave::internal::jacobian_t<Derived, Target> wave::internal::evaluateJacobianAuto(const wave::ExpressionBase<T>&, const TargetDerived&) [with Derived = wave::Rotate<wave::ExpMap<const wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >&>, const wave::Translation<Eigen::Matrix<double, 3, 1> >&>; TargetDerived = wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >; typename std::enable_if<typename wave::internal::traits<T>::UniqueLeaves{}, int>::type <anonymous> = 0; wave::internal::jacobian_t<Derived, Target> = Eigen::Matrix<double, 3, 3>; typename std::remove_const<typename std::remove_reference<decltype (declval<typename wave::internal::traits<WrtDerived, void>::OutputFunctor>()(declval<typename wave::internal::traits<typename std::remove_const<typename std::remove_reference<typename wave::internal::traits<WrtDerived, void>::EvalType>::type>::type, void>::PlainType>()))>::type>::type = wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >; typename std::remove_const<typename std::remove_reference<decltype (declval<typename wave::internal::traits<T, void>::OutputFunctor>()(declval<typename wave::internal::traits<typename std::remove_const<typename std::remove_reference<typename wave::internal::traits<T, void>::EvalType>::type>::type, void>::PlainType>()))>::type>::type = wave::Translation<Eigen::Matrix<double, 3, 1> >; typename wave::internal::traits<typename std::remove_const<typename std::remove_reference<decltype (declval<typename wave::internal::traits<T, void>::OutputFunctor>()(declval<typename wave::internal::traits<typename std::remove_const<typename std::remove_reference<typename wave::internal::traits<T, void>::EvalType>::type>::type, void>::PlainType>()))>::type>::type>::Scalar = double]’
$REPO/extern/wave_geometry/include/wave/geometry/src/core/base/ExpressionBase.hpp:59:46:   required from ‘wave::internal::jacobian_t<Derived, TargetDerived> wave::ExpressionBase<Derived>::jacobian(const TargetDerived&) const [with TargetDerived = wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >; bool TickPrivateBool__LINE__ = true; typename std::enable_if<(TickPrivateBool__LINE__ && wave::internal::is_leaf_expression<TargetDerived>{}), int>::type <anonymous> = 0; Derived = wave::Rotate<wave::ExpMap<const wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >&>, const wave::Translation<Eigen::Matrix<double, 3, 1> >&>; wave::internal::jacobian_t<Derived, TargetDerived> = Eigen::Matrix<double, 3, 3>; typename std::remove_const<typename std::remove_reference<decltype (declval<typename wave::internal::traits<TargetDerived, void>::OutputFunctor>()(declval<typename wave::internal::traits<typename std::remove_const<typename std::remove_reference<typename wave::internal::traits<TargetDerived, void>::EvalType>::type>::type, void>::PlainType>()))>::type>::type = wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >; typename std::remove_const<typename std::remove_reference<decltype (declval<typename wave::internal::traits<T, void>::OutputFunctor>()(declval<typename wave::internal::traits<typename std::remove_const<typename std::remove_reference<typename wave::internal::traits<T, void>::EvalType>::type>::type, void>::PlainType>()))>::type>::type = wave::Translation<Eigen::Matrix<double, 3, 1> >; typename wave::internal::traits<typename std::remove_const<typename std::remove_reference<decltype (declval<typename wave::internal::traits<T, void>::OutputFunctor>()(declval<typename wave::internal::traits<typename std::remove_const<typename std::remove_reference<typename wave::internal::traits<T, void>::EvalType>::type>::type, void>::PlainType>()))>::type>::type>::Scalar = double]’
$REPO/test/gradientsTestutils.h:440:41:   required from ‘Eigen::Matrix<Type, 3, 3> calc_DwxpDw_wave_forward(const Eigen::Quaternion<_Scalar>&, const Eigen::Matrix<Type, 3, 1>&) [with scalar_type = double]’
$REPO/test/gradientsTest.cpp:504:71:   required from here
$REPO/extern/wave_geometry/include/wave/geometry/src/core/functions/TypedJacobianEvaluator.hpp:265:62: error: no matching function for call to ‘wave::internal::TypedJacobianEvaluator<wave::Rotate<wave::ExpMap<const wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >&>, const wave::Translation<Eigen::Matrix<double, 3, 1> >&>, wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >, void>::TypedJacobianEvaluator(<brace-enclosed initializer list>)’
internal::TypedJacobianEvaluator<Derived, TargetDerived> j_eval{v_eval, target};
^~~~~~
$REPO/extern/wave_geometry/include/wave/geometry/src/core/functions/TypedJacobianEvaluator.hpp:178:24: note: candidate: wave::internal::TypedJacobianEvaluator<Derived, Target, typename std::enable_if<(((wave::internal::is_binary_expression<T>::value && (! std::is_same<_From, _To>{})) && wave::internal::contains_same_type<typename wave::internal::traits<T>::LhsDerived, B>::value) && (! wave::internal::contains_same_type<typename wave::internal::traits<T>::RhsDerived, B>::value)), void>::type>::TypedJacobianEvaluator(const wave::internal::Evaluator<Derived>&, const Target&) [with Derived = wave::Rotate<wave::ExpMap<const wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >&>, const wave::Translation<Eigen::Matrix<double, 3, 1> >&>; Target = wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >; typename std::enable_if<(((wave::internal::is_binary_expression<T>::value && (! std::is_same<_From, _To>{})) && wave::internal::contains_same_type<typename wave::internal::traits<T>::LhsDerived, B>::value) && (! wave::internal::contains_same_type<typename wave::internal::traits<T>::RhsDerived, B>::value)), void>::type = void]
WAVE_STRONG_INLINE TypedJacobianEvaluator(const Evaluator<Derived> &evaluator,
^~~~~~~~~~~~~~~~~~~~~~
$REPO/extern/wave_geometry/include/wave/geometry/src/core/functions/TypedJacobianEvaluator.hpp:178:24: note:   no known conversion for argument 1 from ‘const wave::internal::Evaluator<wave::Rotate<wave::ExpMap<const wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >&>&&, const wave::Translation<Eigen::Matrix<double, 3, 1> >&>, void>’ to ‘const wave::internal::Evaluator<wave::Rotate<wave::ExpMap<const wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >&>, const wave::Translation<Eigen::Matrix<double, 3, 1> >&>, void>&’
$REPO/extern/wave_geometry/include/wave/geometry/src/core/functions/TypedJacobianEvaluator.hpp:152:8: note: candidate: wave::internal::TypedJacobianEvaluator<wave::Rotate<wave::ExpMap<const wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >&>, const wave::Translation<Eigen::Matrix<double, 3, 1> >&>, wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >, void>::TypedJacobianEvaluator(const wave::internal::TypedJacobianEvaluator<wave::Rotate<wave::ExpMap<const wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >&>, const wave::Translation<Eigen::Matrix<double, 3, 1> >&>, wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >, void>&)
struct TypedJacobianEvaluator<
^~~~~~~~~~~~~~~~~~~~~~~
  Derived,
~~~~~~~~
  Target,
~~~~~~~
  std::enable_if_t<
~~~~~~~~~~~~~~~~~
    is_binary_expression<Derived>::value && !std::is_same<Derived, Target>{} &&
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    contains_same_type<typename traits<Derived>::LhsDerived, Target>::value &&
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    !contains_same_type<typename traits<Derived>::RhsDerived, Target>::value>> {
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
$REPO/extern/wave_geometry/include/wave/geometry/src/core/functions/TypedJacobianEvaluator.hpp:152:8: note:   candidate expects 1 argument, 2 provided
$REPO/extern/wave_geometry/include/wave/geometry/src/core/functions/TypedJacobianEvaluator.hpp:152:8: note: candidate: wave::internal::TypedJacobianEvaluator<wave::Rotate<wave::ExpMap<const wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >&>, const wave::Translation<Eigen::Matrix<double, 3, 1> >&>, wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >, void>::TypedJacobianEvaluator(wave::internal::TypedJacobianEvaluator<wave::Rotate<wave::ExpMap<const wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >&>, const wave::Translation<Eigen::Matrix<double, 3, 1> >&>, wave::RelativeRotation<Eigen::Matrix<double, 3, 1> >, void>&&)
$REPO/extern/wave_geometry/include/wave/geometry/src/core/functions/TypedJacobianEvaluator.hpp:152:8: note:   candidate expects 1 argument, 2 provided
ERROR: Process failed with exit code 1.
The following products could not be built for configuration release:
gradientsTest

```
However, the following works (reverse-mode?):

```
template<typename scalar_type>
Eigen::Matrix<scalar_type, 3, 3> calc_DwxpDw_wave_reverse(const Eigen::Quaternion<scalar_type>& q,
                                                          const Eigen::Matrix<scalar_type, 3, 1>& p) {
    const wave::RelativeRotation<Eigen::Matrix<scalar_type, 3, 1>> w(Sqis::vec_from_quat(q));
    const wave::Translation<Eigen::Matrix<scalar_type, 3, 1>> t(p);

    const auto [wxp, DwxpDw, DwxpDp] = (wave::exp(w) * t).evalWithJacobians(w, t);
    (void)wxp; (void)DwxpDp;
    return DwxpDw;
}
```

--- a/include/wave/geometry/src/core/functions/TypedJacobianEvaluator.hpp
+++ b/include/wave/geometry/src/core/functions/TypedJacobianEvaluator.hpp
@@ -262,7 +262,9 @@ auto evaluateTypedJacobian(const ExpressionBase<Derived> &expr,
     // Note since we don't return the value, we don't need the user-facing OutputType
     using OutType = eval_output_t<Derived>;
     const auto &v_eval = prepareEvaluatorTo<OutType>(expr.derived());
-    internal::TypedJacobianEvaluator<Derived, TargetDerived> j_eval{v_eval, target};
+    using ExprType = tmp::remove_cr_t<decltype(v_eval.expr)>;
+
+    internal::TypedJacobianEvaluator<ExprType, TargetDerived> j_eval{v_eval, target};
     const auto &result = j_eval.jacobian();
     return result;
 }

--- a/test/rotation_test.cpp
+++ b/test/rotation_test.cpp
@@ -144,7 +144,10 @@ TYPED_TEST_P(RotationTest, rotateVectorByInverse) {
     // As in other tests, CHECK_JACOBIANS checks all Jacobian evaluators, but some
     // high-level glue code that constructs the evaluators isn't checked by that macro.
     // Check it here for a few cases.
-    const auto [p2, J_p2_wrt_r1, J_p2_wrt_p1] = expr.evalWithJacobians(r1, p1);
+    typename TestFixture::PointBAB p2;  // gcc-6 support
+                                        // @todo gcc-7: update to structured binding
+    typename TestFixture::Matrix3 J_p2_wrt_r1, J_p2_wrt_p1;
+    std::tie(p2, J_p2_wrt_r1, J_p2_wrt_p1) = expr.evalWithJacobians(r1, p1);
     const auto J_p2_wrt_r1_single = expr.jacobian(r1);
     const auto J_p2_wrt_p1_single = expr.jacobian(p1);
 

--- a/test/rotation_test.cpp
+++ b/test/rotation_test.cpp
@@ -129,6 +129,30 @@ TYPED_TEST_P(RotationTest, rotateVector) {
     CHECK_JACOBIANS(true, r1 * p1, r1, p1);
 }
 
+TYPED_TEST_P(RotationTest, rotateVectorByInverse) {
+    const auto r1 = TestFixture::LeafAB::Random();
+    const auto p1 = TestFixture::PointAAB::Random();
+    const auto expr = inverse(r1) * p1;
+    const auto p2_single = typename TestFixture::PointBAB{expr};
+
+    const auto eigen_result = (r1.value().matrix().transpose() * p1.value()).eval();
+    EXPECT_APPROX(eigen_result, p2_single.value());
+    CHECK_JACOBIANS(true, expr, r1, p1);
+
+    // Test that calls to expr.jacobian(x) work the same as evalWithJacobians().
+    //
+    // As in other tests, CHECK_JACOBIANS checks all Jacobian evaluators, but some
+    // high-level glue code that constructs the evaluators isn't checked by that macro.
+    // Check it here for a few cases.
+    const auto [p2, J_p2_wrt_r1, J_p2_wrt_p1] = expr.evalWithJacobians(r1, p1);
+    const auto J_p2_wrt_r1_single = expr.jacobian(r1);
+    const auto J_p2_wrt_p1_single = expr.jacobian(p1);
+
+    EXPECT_APPROX(p2_single, p2);
+    EXPECT_APPROX(J_p2_wrt_p1, J_p2_wrt_p1_single);
+    EXPECT_APPROX(J_p2_wrt_r1, J_p2_wrt_r1_single);
+}
+
 TYPED_TEST_P(RotationTest, inverseExpr) {
     const auto r1 = TestFixture::LeafAB::Random();
     const auto r2 = typename TestFixture::LeafBA{inverse(r1)};
@@ -202,6 +226,7 @@ REGISTER_TYPED_TEST_CASE_P(RotationTest,
                            assignViaGetter,
                            getTranslation,
                            rotateVector,
+                           rotateVectorByInverse,
                            inverseExpr,
                            composeWithM,
                            composeWithQ,

